### PR TITLE
adding null integration test

### PR DESCRIPTION
**Pull request by walterddr**
null handling was only tested against FLOAT and DOUBLE peviously
- null value test for more types
- add null integration test

### CODE REVIEW

File: pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java [hunk 1]
```diff
@@ -317,16 +395,17 @@ public void testQueries(Number baseValue, ColumnDataType dataType, boolean nullV
           //(3 rows)
           assertEquals(row[3], 0L);
         } else if ((int) row[keyColumnIdx] == 1) {
-          assertTrue(Math.abs(((Double) row[0]) - 4 * _sumKey1) < 1e-1);
-          assertTrue(Math.abs(((Double) row[1]) - baseValue.doubleValue()) < 1e-1);
+          assertTrue(Math.abs(((Double) row[0]) - 4 * _sumKey1) < PRECISION);
```
Comment (jackjlli) [inline pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java:398]: I saw the value of `PRECISION` constant is `1` in this PR, isn't it greater than `1e-1`?

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java [hunk 2]
```diff
@@ -132,6 +132,9 @@ protected void testQuery(String pinotQuery, String h2Query)
   @AfterClass
   public void tearDown()
       throws Exception {
+    // Setting data table version back
+    DataTableBuilderFactory.setDataTableVersion(DataTableBuilderFactory.DEFAULT_VERSION);
```
Comment (jackjlli) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:136]: Do we need to set the following to the `setUp` method of this integration test class as well?
```
    // Setting data table version to 4
    DataTableBuilderFactory.setDataTableVersion(DataTableFactory.VERSION_4);
```

File: pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java (hunk 1 above)
Comment (walterddr) [reply pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java:398]: yeah this was for integer and long in the future. but yes i can make it 1e-1 for now until we add other types

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java (hunk 2 above)
Comment (walterddr) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:136]: yeah this is a hotfix, multistage test was already setting it in setup but it wasn't set back in teardown i just happen to notice

File: pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java [hunk 3]
```diff
@@ -156,131 +179,186 @@ private void setUp(TableConfig tableConfig, DataType dataType)
...
-            .build();
-    setUp(tableConfig, columnDataType.toDataType());
-    testQueries(baseValue, columnDataType, generateNulls);
+  @DataProvider(name = "numberColumnTypeProvider")
+  public Object[][] provideNumberColumnTypeAndRandomVal() {
+    return new Object[][]{
+        new Object[]{ColumnDataType.FLOAT, RANDOM.nextFloat(), true},
+        new Object[]{ColumnDataType.DOUBLE, RANDOM.nextDouble(), true},
+        new Object[]{ColumnDataType.FLOAT, RANDOM.nextFloat(), false},
+        new Object[]{ColumnDataType.DOUBLE, RANDOM.nextDouble(), false},
```
Comment (walterddr) [inline pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java:238]: @nizarhejazi i was trying to add additional numeric test types here but the assert logic is a bit hard to generalized. please kindly take a look and see if you have better ideas.

File: pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java [hunk 4]
```diff
@@ -156,131 +179,186 @@ private void setUp(TableConfig tableConfig, DataType dataType)
...
+        new Object[]{ColumnDataType.INT, RANDOM.nextInt(), true},
+        new Object[]{ColumnDataType.LONG, RANDOM.nextInt(), true},
+        new Object[]{ColumnDataType.BOOLEAN, RANDOM.nextInt(2), true},
+        new Object[]{ColumnDataType.STRING, RANDOM.nextInt(), true},
+        new Object[]{ColumnDataType.TIMESTAMP, RANDOM.nextInt(), true},
+        new Object[]{ColumnDataType.INT, RANDOM.nextInt(), false},
+        new Object[]{ColumnDataType.LONG, RANDOM.nextInt(), false},
+        new Object[]{ColumnDataType.BOOLEAN, RANDOM.nextInt(2), false},
+        new Object[]{ColumnDataType.STRING, RANDOM.nextInt(), false},
+        new Object[]{ColumnDataType.TIMESTAMP, RANDOM.nextInt(), false},
```
Comment (walterddr) [inline pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java:254]: this list is not at all a full coverage, but good to add STRING BOOLEAN and TIMESTAMP as representatives. can add more later

File: pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java [hunk 5]
```diff
@@ -156,131 +179,186 @@ private void setUp(TableConfig tableConfig, DataType dataType)
...
+      }
+      // The default null ordering is 'NULLS LAST'.
+      if (nullValuesExist) {
+        Object[] row = rows.get(rows.size() - 1);
+        assertEquals(row[0], 2000L);
+        assertNull(row[1]);
+      }
+    }
+    // TODO fix String null value handling
+    if (dataType != ColumnDataType.STRING) {
```
Comment (walterddr) [inline pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java:330]: looks like this query doesn't work with STRING type. might be we can take another look @nizarhejazi

**Comment by nizarhejazi** [comment]
@walterddr NullHandling support was tested against all data types. Please check the following files:
- NullEnabledQueriesTest (Dict and non-dict Float and Double types).
- BooleanNullEnabledQueriesTest (Dict and non-dict Boolean type).
- BigDecimalQueriesTest (Dict and non-dict type).
- AllNullQueriesTest (Dict and non-dict Long, Float, Double, Int, String and BigDecimal types).

File: pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java (hunk 4 above)
Comment (nizarhejazi) [reply pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java:254]: @walterddr NullHandling support was tested against all data types. Please check the following files:

NullEnabledQueriesTest (Dict and non-dict Float and Double types).
BooleanNullEnabledQueriesTest (Dict and non-dict Boolean type).
BigDecimalQueriesTest (Dict and non-dict type).
AllNullQueriesTest (Dict and non-dict Long, Float, Double, Int, String and BigDecimal types).

File: pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java (hunk 5 above)
Comment (nizarhejazi) [reply pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java:330]: @walterddr There is a reason this query does not work w/ String, right? We cannot pass String column to aggregation functions (Count, Min, Max, Avg, etc.) and we cannot use inequality operators w/ Strings.

File: pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java (hunk 4 above)
Comment (walterddr) [reply pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java:254]: Ah I see thanks for the info. I wasn't aware of the other tests

File: pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java (hunk 5 above)
Comment (walterddr) [reply pinot-core/src/test/java/org/apache/pinot/queries/NullEnabledQueriesTest.java:330]: Oh I see. So we are not using comparison for min/max and null checker for count. But actually fallback to the numeric handling. Is this correct?

**Comment by walterddr** [comment]
looks like as @nizarhejazi mentioned the issue is not on null handling (e.g. select out null column) but on function interpretation of null values. closing this and creating a new PR for fixing the issue

**Pull request closed without merging**